[docs] migrate Roles page (#215)
moodle/devdocs: docs/apis/subsystems/roles.md

Permissions control possible user actions within Moodle (eg delete discussions, add activities etc)

Roles can be applied to users in a **context** (eg assign Fred as a teacher in a particular course)

### Context

Here are the possible contexts, listed from the most general to the most specific.

| CONTEXT NAME        | CONTEXT AREA       | CONTEXT LEVEL |
|---------------------|--------------------|---------------|
| `CONTEXT_SYSTEM`    | the whole site     | 10            |
| `CONTEXT_USER`      | another user       | 30            |
| `CONTEXT_COURSECAT` | a course category  | 40            |
| `CONTEXT_COURSE`    | a course           | 50            |
| `CONTEXT_MODULE`    | an activity module | 70            |
| `CONTEXT_BLOCK`     | a block            | 80            |

An authorized user will be able to assign an arbitrary number of roles to each user in any context.

See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.

### Capabilities

Capabilities can have the following permissions:

1. `CAP_INHERIT`
2. `CAP_ALLOW`
3. `CAP_PREVENT`
4. `CAP_PROHIBIT`

> If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.

Reviewer: ```suggestion
:::info

If no permission is defined for a capability in a role, then the permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.

:::
```

### Capability conflicts

> Since the capabilities in each role could be different, there could be conflict in capabilities.

Reviewer: ```suggestion
Since the capabilities in each role could be different, there can be conflicts in capabilities.
```

> If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".

Reviewer: This could do with some more examples:

- User is a teacher and a student
- The `moodle/site:accessallgroups` capability granted to the teacher, but is prevented for the student
- The `moodle/site:accessallgroups` capability granted to the teacher, but is prevented for the student in the category
 - The `moodle/site:accessallgroups` capability granted to the teacher, but is prohibitted for the student in the category
 - The `moodle/site:accessallgroups` capability granted to the teacher in the category, but is prevented for the student in the course

## Programming Interface

Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.

- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)

> 

Reviewer: ```suggestion
```

> - Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
>     those predefined roles using **role archetype:**

Reviewer: ```suggestion
- Moodle comes with a list of predefined roles, including a Student, Teacher, Non-editing teacher, and Course Manager role. Each of these roles are based on a **role archetype**, which acts as a template for roles. Any custom role created by the site administrator can also _choose_ to follow one of these role archetypes. When a plugin defines a new capability, it may specify how it would expect to be applied within these role archetypes, and these are applied to any role which follows this archetype. For example, if you create a new activity module with named _mod_example_, with a capability 'mod/example:view', you may specify that the `teacher`, and `editingteacher` archetypes are granted the capability with the allow permission. Any role which is based on these archetypes will be granted this capability with the 'allow' permission.
```

> ```php
> /**
>  * Returns array of all role archetypes.
>  *
>  * @return array
>  */
> function get_role_archetypes() {
>     return array(
>         'manager'        => 'manager',
>         'coursecreator'  => 'coursecreator',
>         'editingteacher' => 'editingteacher',
>         'teacher'        => 'teacher',
>         'student'        => 'student',
>         'guest'          => 'guest',
>         'user'           => 'user',
>         'frontpage'      => 'frontpage'
>     );
> }
> ```

Reviewer: I'd query whether we should put this in in this format. You could instead do something like:

```suggestion

The role archetypes do not change often, and are currently defined as:

- **manager** - A course category role used to manage courses without being directly enrolled in them
- **coursecreator** - ...
- ...

<details>
<summary>Fetching a list of the role archetypes programatically</summary>

In some rare situations you may need to fetch a list of available role archetypes. You can do so using the `get_role_archetypes()` function, for example:

```php title="Fetching a list of role archetypes"
$archetypes = get_role_archetypes();
```

</details>

```

> - On each page you need to find the context the user is working in, using the `get_context_instance()` function.  For example, in the forum module:

Reviewer: `get_context_instance()` is depreceated.

Better to use one of:

- `context::instance_by_id()`
- `context_[type]::instance($typeid)`

```php
$context = get_context_instance(CONTEXT_MODULE, $cm->id);
```

or at the course level:

```php
$context = get_context_instance(CONTEXT_COURSE, $id);
```

> - Then, whenever you want to check which role has a given capability do something like this:
> 
> ```php
> list($roleids) = get_roles_with_cap_in_context($context, 'moodle/course:manageactivities');
> ```

Reviewer: We should discourage this. It's one of the things that people think they need but almost never actually need. It leads to questions like "How do I know who my teachers are", and "How do I make this work for teachers only".

Perhaps put it in a `<details>` section so that it is clear that it is not a part of the norm:

```suggestion
<details>
<summary>Fetching roles and users who hold a capability</summary>

Moodle has a flexible and detailed capability system which allows administrators to define many similar roles for different purposes. It is quite common to have multiple teacher-like roles but need to restrict their access depending on their usage.

For example in a University setting you may have a lecturer who presents the course materials, and then a number of Ph.D students who lead smaller groups of students in labs, seminars, and workshops. These roles may both be considered a form of teacher, but they will have different permissions to suit their needs.

As a result we strongly discourage that you think in terms of which roles or users hold a capability, but rather whether a specific user holds a capability.

:::danger

There are some situations where you _do_ need to get a list of roles with a capability in a specific context, but these are very rare. You can do so using the `get_roles_with_cap_in_context()` function:

```php title="Fetching a list of roles which old a capability in the specified context"
[$roleids] = get_roles_with_cap_in_context($context, 'moodle/course:manageactivities');
```

:::

</detail>
```

- If you want to get a list of roles that a give user has you can call

```php
$ras = get_user_roles($context, $user->id);
```

- You can assign a role to a user by calling something like

```php
role_assign($roleid, $userid, $context->id);
```

> 

Reviewer: Again, this is not something we should encourage.

Unless you are writing an enrolment plugin it is highly unlikely that you will need these features, and when you do, you will probably want the extra parameters to them - that is the `$component, $itemid` parms to `role_assign()`.

I'd maybe add a warning again, especially to `get_user_roles()`, but also mention that `role_assign()` is typically only used for Enrolment plugins (link to their docs).

## See also

> - [Hardening new Roles system](https://docs.moodle.org/dev/Hardening_new_Roles_system)

Reviewer: Let's remove the word 'new' - it's been over 10 years
```suggestion
- [Hardening the Roles system](https://docs.moodle.org/dev/Hardening_new_Roles_system)
```

Ideally we should merge that document into this one.

> - Using Moodle course:

Reviewer: ```suggestion
- Relevant discussions on [moodle.org](https://moodle.org/):
```

  - [Roles and Capabilities forum](http://moodle.org/mod/forum/view.php?f=941)
  - Key discussions at Using Moodle forums:

>     - [Roles and Permissions architecture](http://moodle.org/mod/forum/discuss.php?d=38788)

Reviewer: Not worth mentioning. 2006 from Moodle 1.7
```suggestion
```

>     - [An example of admin roles as set in the database](http://moodle.org/mod/forum/discuss.php?d=56302#256313)

Reviewer: This relates to Moodle 1.7b from 2006. Everything is diferent.
```suggestion
```

---
title: Roles API
tags:
  - Roles
  - Interfaces
  - Capabilities
---

> ## Definitions

Reviewer: We typically add a brief introduction to the topic at the start of every page rather than jumping into subtitles.

I'd suggest addign a paragraph to the top here to explain what the Roles API is for.

> - A **role** is an identifier of the user's status in some context. For example: Teacher, Student and Forum moderator are examples of roles.
> 
> - A **capability** is a description of some particular Moodle feature. Capabilities are associated with roles. For example, `mod/forum:replypost` is a capability.
> 
> - A **permission** is some value that is assigned for a capability for a particular role.  For example, allow or prevent.
> 
> - A **context** is a "space" in the Moodle, such as courses, activity modules, blocks etc.

Reviewer: This needs tidying up and improving. The examples aren't very clear.

```suggestion
- A **context** is an area in Moodle. There are several context types for:
  - the whole site
  - a user
  - a course category
  - a course
  - an activity
  - a block
- A **role** is an identifier of the user's status in some context. Teacher, Student and Forum moderator are all examples of possible roles.
- A **capability** is a description of some particular Moodle feature. Capabilities belong to a component. and are assigned to roles. For example, `mod/forum:replypost` is a capability.
- A **permission** is the level in which a capability is applied to a role. Example values include:
  - Allow
  - Prevent
  - Prohibit
  - Inherit
```

> ## The system prior to v1.7
> 
> In versions prior to v1.7, Moodle uses a fixed set of roles i.e. primary admin, admins, course creators, editing teachers, non-editing teachers, students, and guests. For each role, the capability or actions that they can perform are fixed. For example, the role student allows the user to submit an assignment, but doesn't allow the user to browse/edit other users' work. By using this setup we limit ourselves to a rather rigid set of capabilities for each role. If we want, say a particular student or group to be able to mark assignments in a particular course, we can't do that without giving these users teacher privileges.
> 
> ## The new roles and capability system
> 
> With v1.7 and greater, Moodle introduces a roles and capabilities system.
> 

Reviewer: Remove this. Roles haven't changed since 2.0. 1.7 is ancient - the final version came out in January 2009, so 13 years since support _ended, and 16 since it stareted

```suggestion
```

4. `CAP_PROHIBIT`

If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.

### Checking for capabilities

You can use the [has_capability ()](https://docs.moodle.org/dev/Roles_and_modules#has_capability_.28.24capability.2C_.24contextid.2C_.24kill.29) function for checking for capabilities in your code.

### Capability conflicts

Since the capabilities in each role could be different, there could be conflict in capabilities. This is resolved using different methods in versions below Moodle 2.0.

If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".

> ### Capability-locality changes in v1.9
> 
> When resolving conflicts between "allow" and "prevent" in v1.7 and v1.8 the locality of the capability is taken into account, although with less weight than the locality of the role assignment. In v1.9 the process has been reworked, we consider locality of the role assignment when dealing with allow/prevent conflicts, but ignore where the capability has been defined (as long as it applies to the context).
> 
> Allow and prevent will cancel each other out if set for the same capability at the same context level. If this happens, we refer to the previous context level to determine the permission for the capability. This may sound more complex than it really is in practice.  The upshot is that the system can be flexible enough to allow pretty much any combination of permissions.
> 
> For example - for the situation where there is
> 
> - two roles assigned to one user in a context (eg student role and teacher role)
> - one override on one of those roles ( eg student role), at the same context
> 
> then
> 
> - **1.7/1.8** The override used to **always** win over the settings from the combined roles.
> 
> - **1.9** The override is combined with the student role first. and then the roles are combined. This is more logical, as the override is acting like a modification to the student role.
> 
> For more details, see [MDL-11218](http://tracker.moodle.org/browse/MDL-11218)
> 

Reviewer: ```suggestion
```

### New permissions evaluation in 2.0

> The permission evaluation changes introduced in 1.9 were too complex for majority of uses, the permission evaluation was finally simplified in Moodle 2.0 - see [New permissions evaluation in 2.0](https://docs.moodle.org/dev/New_permissions_evaluation_in_2.0).

Reviewer: Perhaps we should move this link to be our new ROles documentation since it's describing what we do with roles in 20+?

Author: Nothing on useful on New_permissions_evaluation_in_2.0, so just removing

> ## Upgrading from 1.6
> 
> A smooth upgrade will be provided with 1.7. The existing roles (admin, teacher, student, etc), and the existing capabilities will be automatically retained.  This is done by creating default roles at site/course levels, and assigning the current users to these roles accordingly. The default roles will have default capabilities associated with them, mirroring what we have  in 1.6.   With no modifications, Moodle will operate exactly the same before and after the upgrade.
> 

Reviewer: ```suggestion
```

    1. `mod/glossary:comment` - comment on an entry
    1. `mod/glossary:managecomments` - edit, delete comments
    1. `mod/glossary:import` - import glossaries
    1. `mod/glossary:export` - export glossaries
    1. `mod/glossary:approve` - approve glossaries
    1. `mod/glossary:rate` - rates glossary
    1. `mod/glossary:viewrating` - view ratings
8. Hotpot
    1. `mod/hotpot:attempt` - attempt a hotpot
    1. `mod/hotpot:viewreport` - review and view reports
    1. `mod/hotpot:grade` - (grade? and) regrade
    1. `mod/hotpot:deleteattempt` - deletes attempts
9. Label
    1. none
10. Lams
    1. `mod/lams:participate` - original student
    1. `mod/lams:manage` - original teacher
11. Lesson
    1. `mod/lesson:edit` - add and edit pages
    1. `mod/lesson:manage` - view student attempts
12. #Quiz
     1. `mod/quiz:grade` - comment, override grade, manual grade
     1. `mod/quiz:preview` - previews the quiz
     1. `mod/quiz:viewreports` - view quiz result reports
     1. `mod/quiz:manage` - add/delete/move (up or down) questions for a quiz
     1. `mod/quiz:attempt` - attempt the quiz--[Tim Hunt](https://docs.moodle.org/dev/User:Tim_Hunt)
13. Resource
14. Scorm

>     1. `mod/scorm:viewgrades`
> 15. Survey
>     1. `mod/survey:download` - downloads server result
>     1. `mod/survey:participate` - participate/ do survey
>     1. `mod/survey:readresponses` - read all user's response
> 16. Wiki
>     1. `mod/wiki:participate` - original student, meaning depends on type and course setting
>     1. `mod/wiki:manage` - original teacher, manages assigned group; `moodle/site:accessallgroups` is needed to manage all groups
>     1. (Waiting on new wiki)
> 17. Workshop
>     1. `mod/workshop:participate` - original student, allows user to submit and assess
>     1. `mod/workshop:manage` - original teacher, user can manage others
>     1. (Waiting on new Workshop)
> 
> ### Enrolment-level Capabilities
> 
> The naming convention for capabilities that are specific to enrolment is `enrol/enrol_name:capability`. The enrolment capabilities are defined in `enrol/enrol_name/db/access.php`.
> 
> 1. Authorize.net Payment Gateway
>    1. `enrol/authorize:managepayments` - manage user payments, capture, void, refund, delete etc.
> 
>    1. Blocks
> 1. `activity_modules`
>    1. None
> 1. `admin`
> 1. `admin_2`
> 1. `admin_bookmarks`
> 1. `blog_menu`
> 1. `blog_tags`
> 1. `calendar_month`
> 1. `calendar_upcoming`
> 1. `course_list`
> 1. `course_summary`
> 1. `glossary_random`
> 1. `html`
> 1. `loancalc`
> 1. `login`
> 1. `messages`
> 1. `news_items`
> 1. `online_users`
> 1. `participants`
> 1. `quiz_results`
> 1. `recent_activity`
> 1. `rss_client`
>     1. `block/rss_client:createprivatefeeds`
>     2. `block/rss_client:createsharedfeeds`
>     3. `block/rss_client:manageownfeeds`
>     4. `block/rss_client:manageanyfeeds`
> 1. `search`
> 1. `search_forums`
> 1. `section_links`
> 1. `site_main_menu`
> 1. `social_activities`
> 
> ### Questions
> 
> I am adding question categories here because they seem to have been forgotten in the whole scheme of things since having been removed from the quiz module itself. I've made a suggestion on how these could be handled in [MDL-6118](https://tracker.moodle.org/browse/MDL-6118).
> 
> : See [this forum thread](http://moodle.org/mod/forum/discuss.php?d=51143) for a discussion about the current problems wth publishing question categories.[Tim Hunt](https://docs.moodle.org/dev/User:Tim_Hunt) 18:50, 8 August 2006 (WST)
> 

Reviewer: Remove this. It's irrelevan tand very out-of-date.

Better to give people the information they need to find this out for themslves

```suggestion
```

### Visitor

This would be a role whereby one could allow a visitor to visit one's classroom. This might be a colleague interested in seeing your course, or a journalist who might be writing an article about one's site. They should not be able to see the names of any students anywhere (eg recent activity, forum posts) for privacy reasons. They should be able to try out things like quizzes, and lessons but no grades would be recorded (like in teacher preview mode). They would not be able to participate in choices and forums but could view them. It would be read only in a way like former-student role below but without access to a particular student's records that former student role would grant.

### Guest Speaker

This role would be similar to the Visitor role above, but would allow seeing student names, and also allow both reading and posting to a specific forum or forums. We often have "guest speakers" who read and respond to student forum posts. Right now we have to add them as students, which isn't ideal.

> ### Former Student
> 
> This role would be of particular use for courses with rolling enrollments. This role would be one where a student had completed all of the requirements of a course (ie. assignments, quizzes etc.) but wished to have continued access to the course material for review or consultation. The key factor is that one would give access to the completed student to the notes he read, his work and the teacher's comments on it, but he would not be allowed to do anything that would take up the teacher's time. In other words, a sort-of read-only access to the course. How forums, which might contain pertinent information and would continue to grow, would be handled is a question. Perhaps the student would be shown only what was in the forums at the time he completed the course. He would not be allowed to see any new posts or add any himself. Same thing for database and glossary entries. In other words, a snapshot of the course at the time his regular enrollment ended. He shouldn't be able to see the names or profiles of any newly enrolled students for privacy reasons-hence the restrictions on forum access. One issue that would have to be dealt with would be changes to existing modules-such as resources. Does the student get access to the module as it was or as it is? We have no versioning of resources in Moodle so this would be a problem. What about a teacher changing a quiz question so that the answer is different? What would a former student see?
> 
> ### Alumnus
> 
> An ALUMNUS should be able to search for all other ALUMNI of the school, interact with them and be enrolled in a separate course - which is like a META course with all the content of his learning and interaction - as well as capabilities to be a part of this ALUMNI only course.  All the teachers of courses during school years should automatically be a part of the ALUMNI course .. which means when an ALUMNUS is enrolled in a course, the original teachers of all his courses get enrolled ?  --[Anil Sharma](https://docs.moodle.org/dev/index.php?title=User:Anil_Sharma&action=edit&redlink=1) 20:54, 15 July 2006 (WST)
> 
> ### Librarian
> 
> Reference Librarians have an active role in most of the courses taught at some schools such as Earlham College (with Bibliographic Instruction). The Librarian role within Moodle could encompass default read access to all courses (unless prohibited by course teacher) and read access to all components of the course unless access is barred (again by teacher). The Librarians would also perhaps have a block called perhaps Reference Services or Reference Desk with write access where they could deposit resources. Also this block might have a chat applet whereby enrolled students could chat to the Reference Librarian on duty about their bibliographic research needs.
> 
> In schools there is often a book review system. This may be covered by the lending system database but may not in which case a librarian may need to have a course area they can create a database template to handle the reviews in which case they may have a normal teacher style role? Off topic but course an integration with common schools database systems would be great.
> 
> ### Teacher
> 
> Teachers should have read access to other Teacher's courses unless explicitly prohibited. They should be able to set parts of their own course to be totally private (perhaps even to admin?). Just as each activity can currently be set to have group access, each activity could have a permissions field. Teachers could set default permissions for all activities on their course (eg they might disallow Librarian access for example) and then change the access permission for an individual activity.
> 
> I think that what is needed is a simple hierarchy of permissions and levels of granularity.
> 
> I would take issue with "teachers should have read access to other teacher's courses unless explicitly prohibited." This is a violation of the students' privacy as how they perform and what they do in one class isn't the business of another teacher. Moreover, in the real world a teacher wouldn't suddenly go sit in on a colleague's class without asking permission first. I would not have appreciated such an invasion of privacy as either a teacher or a student. It could be an option, but shouldn't be default.--[N Hansen](https://docs.moodle.org/dev/User:N_Hansen) 19:54, 12 June 2006 (WST)
> 
> ### Community Education Tutors/Trainers
> 
> Teachers may be community adult education trainers making use of a school moodle so must only have access to their courses unless given access elsewhere. They would not necessarily get the default teacher privileges.
> 
> ### Secretary/Student Worker
> 
> <!-- cspell:ignore FERPA -->
> We often have faculty who want their departmental secretary or student worker to scan and upload files and perhaps create resources. Currently they have to be given teacher access to the course. This is dangerous from a FERPA standpoint since they could easily get access to grades.
> 
> ### Teaching Assistant
> 
> Our Faculty frequently have undergraduate students acting as Teaching Assistants. These students need to be able to add resources, create assignments, and possibly grade assignments. However, due to FERPA they cannot have access to other students' overall grade information. I think the requirements here are slightly different than those of Secretary/Student Worker
> 
> ### Student - FERPA rights
> 
> A student that has asserted their FERPA rights to non-disclosure.  Typically includes not publishing their name
> in any public place.  Could include this student only being seen with an "alias" within course spaces.  Is this an attribute rather
> than a role?
> 
> ### Help Desk
> 
> Help desk agents that have read access for the purposes of troubleshooting.  Some care in placing this role within a hierarchy
> of inheritance is needed, full access will be problematic with FERPA.
> 
> ### Admin - Category based
> 
> Basically a person in between full Admin and Creator that has the permissions of an Admin but only with respect to courses and students. Currently a Creator has permissions site-wide which does not always meet the requirements of a given organisation (e.g. Department A may not be happy that a person from Department B can create/modify courses within Department A's area). The ability to designate a Creator within a specific category would allow areas to be set up for a faculty/department/organisation and allow the Admin for that area to create/delete courses, upload users, add site-wide entries to the calendar etc.
> 
> ### Process Roles
> 
> organising the learning process for a group you wish to have the choice to place students in different roles: examples of this are:
> <!-- cspell:ignore rolegame -->
> 
> - Give a student the role of forum-moderator with edit and chunk-rights
> - Give students different roles & rights in a web quest design (and change these roles next week
> - Give students different resources, depending on their roles in a rolegame/simulation
> - Give a student the rights to create the section content of next week (and only that week..)
> 

Reviewer: Irrelevant
```suggestion
```
